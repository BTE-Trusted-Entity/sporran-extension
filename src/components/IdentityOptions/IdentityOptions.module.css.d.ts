export const __esModule: true;
export const wrapper: string;
export const toggle: string;
