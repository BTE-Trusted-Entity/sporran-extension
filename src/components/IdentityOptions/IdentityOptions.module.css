.wrapper {
  composes: wrapper from '../../components/Menu/Menu.module.css';
  display: inline-flex;
  margin-left: 0.5rem;
}

.toggle {
  composes: toggle from '../../components/Menu/Menu.module.css';
  composes: buttonIcon from '../../components/Button/Button.module.css';
  width: 1.625rem;
  height: 1.625rem;
  background-size: 0.75rem;
  background-image: url('../../components/images/chevron-down.svg');
}

.toggle[aria-expanded='true'] {
  transform: rotate(180deg);
}
