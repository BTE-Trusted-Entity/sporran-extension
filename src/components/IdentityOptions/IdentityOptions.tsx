import { useCallback } from 'react';
import { Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';
import useDropdownMenu from 'react-accessible-dropdown-menu-hook';

import * as menuStyles from '../Menu/Menu.module.css';
import * as styles from './IdentityOptions.module.css';

import { Identity } from '../../utilities/identities/types';
import { generatePath, paths } from '../../views/paths';

interface Props {
  identity: Identity;
  onEdit: () => void;
}

export function IdentityOptions({ identity, onEdit }: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const { address } = identity;

  const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(3);

  const handleClick = useCallback(() => {
    onEdit();
    setIsOpen(false);
  }, [onEdit, setIsOpen]);

  return (
    <div className={styles.wrapper}>
      <button
        {...buttonProps}
        type="button"
        title={t('component_IdentityOptions_label')}
        aria-label={t('component_IdentityOptions_label')}
        className={styles.toggle}
      />

      {isOpen && (
        <div className={menuStyles.dropdown} role="menu">
          <h4 className={menuStyles.heading}>
            {t('component_IdentityOptions_heading')}
          </h4>

          <ul className={menuStyles.list}>
            <li className={menuStyles.listItem}>
              <button
                type="button"
                {...(itemProps[0] as Record<string, unknown>)}
                onClick={handleClick}
                className={menuStyles.listButton}
              >
                {t('component_IdentityOptions_edit')}
              </button>
            </li>
            <li className={menuStyles.listItem}>
              <Link
                to={generatePath(paths.identity.remove, { address })}
                {...itemProps[1]}
              >
                {t('component_IdentityOptions_forget')}
              </Link>
            </li>
            <li className={menuStyles.listItem}>
              <Link
                to={generatePath(paths.identity.reset.start, { address })}
                {...itemProps[2]}
              >
                {t('component_IdentityOptions_reset_password')}
              </Link>
            </li>
          </ul>
        </div>
      )}
    </div>
  );
}
