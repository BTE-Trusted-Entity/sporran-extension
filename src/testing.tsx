export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
