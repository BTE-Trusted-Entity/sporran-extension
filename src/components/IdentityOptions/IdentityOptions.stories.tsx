import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import * as menuStyles from '../Menu/Menu.module.css';

import { identitiesMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { IdentityOptions } from './IdentityOptions';

export default {
  title: 'Components/IdentityOptions',
  component: IdentityOptions,
  decorators: [
    (Story) => (
      <div className={menuStyles.wrapper} style={{ float: 'right' }}>
        <Story />
      </div>
    ),
  ],
} as Meta;

export function Template(): JSX.Element {
  return (
    <IdentityOptions
      onEdit={action('onEdit')}
      identity={
        identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
      }
    />
  );
}
