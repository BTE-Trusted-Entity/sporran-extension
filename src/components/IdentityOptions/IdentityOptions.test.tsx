import userEvent from '@testing-library/user-event';

import { identitiesMock, render, screen } from '../../testing/testing';

import { IdentityOptions } from './IdentityOptions';

const onEdit = jest.fn();

const props = {
  identity: identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'],
  onEdit,
};

describe('AddIdentity', () => {
  it('should render', async () => {
    const { container } = render(<IdentityOptions {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('menu should be visible when menu button clicked', async () => {
    const { container } = render(<IdentityOptions {...props} />);
    const openMenuButton = await screen.findByLabelText('Identity options');
    expect(openMenuButton).toHaveAttribute('aria-expanded', 'false');
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();

    await userEvent.click(openMenuButton);

    expect(await screen.findByRole('menu')).toBeInTheDocument();
    expect(openMenuButton).toHaveAttribute('aria-expanded', 'true');

    expect(container).toMatchSnapshot();
  });
});
