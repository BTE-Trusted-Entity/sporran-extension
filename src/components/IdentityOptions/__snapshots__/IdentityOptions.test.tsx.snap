// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddIdentity menu should be visible when menu button clicked 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="Identity options"
      class="toggle"
      role="button"
      tabindex="0"
      title="Identity options"
      type="button"
    />
    <div
      class="dropdown"
      role="menu"
    >
      <h4
        class="heading"
      >
        Identity
      </h4>
      <ul
        class="list"
      >
        <li
          class="listItem"
        >
          <button
            class="listButton"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            Edit Identity Name
          </button>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/remove"
            role="menuitem"
            tabindex="-1"
          >
            Remove Identity
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/reset"
            role="menuitem"
            tabindex="-1"
          >
            Reset Password
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`AddIdentity should render 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Identity options"
      class="toggle"
      role="button"
      tabindex="0"
      title="Identity options"
      type="button"
    />
  </div>
</div>
`;
